src/components/ha-icon-picker.ts

```diff
@@ -87,48 +130,57 @@ export class HaIconPicker extends LitElement {
     `;
   }
 
-  private async _openedChanged(ev: PolymerChangedEvent<boolean>) {
-    this._opened = ev.detail.value;
-    if (this._opened && !iconLoaded) {
-      const iconList = await import("../../build/mdi/iconList.json");
+  // Filter can take a significant chunk of frame (up to 3-5 ms)
+  private _filterIcons = memoizeOne(
+    (filter: string, iconItems: IconItem[] = ICONS) => {
+      if (!filter) {
+        return iconItems;
+      }
 
-      iconItems = iconList.default.map((icon) => ({
-        icon: `mdi:${icon.name}`,
-        keywords: icon.keywords,
-      }));
-      iconLoaded = true;
+      const filteredItems: RankedIcon[] = [];
+      const addIcon = (icon: string, rank: number) =>
+        filteredItems.push({ icon, rank });
+
+      // Filter and rank such that exact matches rank higher, and prefer icon name matches over keywords
+      for (const item of iconItems) {
+        if (item.parts.has(filter)) {
+          addIcon(item.icon, 1);
```

Comment:
I think adding the size of `parts` to the ranking might make it a bit better.

The when you search `tree`, `mdi:tree` will be on top and not `mdi:chart-tree` (or whatever)

Reply:
Yeah that's a good idea.  You'd have to loop manually over `parts` and set the rank to something like `indexFound/item.parts.size`.

I'm afraid to slow it down any further though because the freezing is clearly affected by the timing.  If I play around with the picker enough in a session, I can actually still get Chrome to freeze, but it's not consistent and much much more difficult.  There's definitely an issue in the vaadin code.